train.py

```diff
@@ -169,41 +169,51 @@ def evaluate(model_engine, eval_dataloaders, tb_writer, step, eval_gradient_accu
         tb_writer.add_scalar('eval/eval_time_sec', duration, step)
     return sum(loss) / len(loss) if len(loss) > 0 else None
 
-
-def apply_max_norm_regularization(model, config):
-    # modifed from https://github.com/kohya-ss/sd-scripts/blob/main/networks/lora.py
-    A_keys = []
-    B_keys = []
+def apply_lora_norm_regularization(model, config, lr):
+    # max-norm stuff modified from https://github.com/kohya-ss/sd-scripts/blob/main/networks/lora.py
     norms = []
     keys_scaled = 0
-    lora_scale = config['lora_alpha'] / config['lora_rank']
 
-    state_dict = model.state_dict()
-    for key in state_dict.keys():
-        if 'lora_A' in key:
-            A_keys.append(key)
-            B_keys.append(key.replace('lora_A', 'lora_B'))
-
-    for i in range(len(A_keys)):
-        A = state_dict[A_keys[i]]
-        B = state_dict[B_keys[i]]
-        W = B @ A
-        W *= lora_scale
-
-        if 'scale_weight_norms' in config:
-            max_norm = config['scale_weight_norms']
-            norm = W.norm().clamp(min=max_norm / 2)
-            desired = torch.clamp(norm, max=max_norm)
-            ratio = desired.cpu() / norm.cpu()
-            sqrt_ratio = ratio**0.5
-            if ratio != 1:
-                keys_scaled += 1
-                state_dict[A_keys[i]] *= sqrt_ratio
-                state_dict[B_keys[i]] *= sqrt_ratio
-        else:
-            ratio = 1.0
-        scalednorm = W.norm() * ratio
-        norms.append(scalednorm.item())
+    lora_scale = config['lora_alpha'] / config['lora_rank']
+    weight_decay = config.get('lora_weight_decay', 0.0)
+    max_norm = config.get('lora_max_norm', 0.0)
+
+    for name, param in model.named_parameters():
+        if 'lora_A' in name:
+            A = param
+            B_name = name.replace('lora_A', 'lora_B')
+            B = next(p for n, p in model.named_parameters() if n == B_name)
+            
+            with torch.no_grad():
+                W = lora_scale * (B @ A)
+
+                # Apply decoupled weight decay (indirectly) to the composite matrix BA if specified
+                # Using L = λ⋅½||W||_F²:
+                #    ∂L/∂W = λ⋅W, as ∂(½||W||_F²)/∂W = W
+                #    ∂L/∂A = s⋅Bᵗ(∂L/∂W) = λ⋅s⋅BᵗW
+                #    ∂L/∂B = s⋅(∂L/∂W)Aᵗ = λ⋅s⋅WAᵗ
+                if weight_decay > 0:                      
+                    grad_A = weight_decay * lora_scale * (B.t() @ W)
+                    grad_B = weight_decay * lora_scale * (W @ A.t())
+                    A -= lr * grad_A
```

Comment:
This should work because it's an in-place operation on a nn.Parameter. I think it's normal that there is some confusion, as "A -= delta" works, but "A = A - delta" does not. Because of that, I think we should change this to be more explicit. Use "A.sub_(delta)" (same with B), and also for the code below "A.mul_(sqrt_ratio)".

Reply:
Yeah, I agree - It's changed now.